src/planwise/component/sources.clj

```diff
@@ -13,15 +16,69 @@
   [store]
   (get-in store [:db :spec]))
 
+(defn- csv-data->maps
+  [csv-data]
+  (map zipmap
+       (->> (first csv-data)
+            (map keyword)
+            repeat)
+       (rest csv-data)))
+
 ;; ----------------------------------------------------------------------
 ;; Service definition
 
 (defn list-sources
   [store owner-id]
   (db-list-sources (get-db store) {:owner-id owner-id}))
 
+(defn create-source-set
+  [store name unit owner-id]
+  (db-create-source-set! (get-db store) {:name name
+                                         :unit unit
+                                         :owner-id owner-id}))
+
+(defn create-source
+  [store source]
+  (println source)
+  (db-create-source! (get-db store) source))
+
+(defn get-source-set
+  [store id owner-id]
+  (db-find-source-set (get-db store) {:id id
+                                      :owner-id owner-id}))
+
+(defn import-source
+  [store set-id csv-source-data]
+  (let [source {:set-id set-id
+                :name (:name csv-source-data)
+                :type (:type csv-source-data)
+                :lat  (Double. (:lat csv-source-data))
+                :lon  (Double. (:lon csv-source-data))
+                :quantity (Integer. (:unsatisfied csv-source-data))}]
+    (create-source store source)))
+
+(defn import-sources
+  [store set-id csv-file]
+  (let [reader (io/reader csv-file)
+        csv-data (csv-data->maps (csv/read-csv reader))]
+    (doall (map #(import-source store set-id %) csv-data))
+    set-id))
+
 (defn import-from-csv
-  [store {:keys [name owner-id]} csv-file])
+  [store {:keys [name unit owner-id]} csv-file]
+  (jdbc/with-db-transaction [tx (get-db store)]
+    (let [tx-store                  (assoc-in store [:db :spec] tx)
+          tx-create-source-set      (fn [name unit]
+                                      (let [result (create-source-set tx-store name unit owner-id)]
+                                        (:id result)))
+          tx-import-sources-from    (fn [set-id csv-file] (import-sources tx-store set-id csv-file))
+          tx-get-created-source-set (fn [set-id] (get-source-set tx-store set-id owner-id))]
+      (-> (tx-create-source-set name unit)
+          (tx-import-sources-from csv-file)
+          (tx-get-created-source-set)))))
```

Comment:
This is a bit convoluted. I think a more explicit implementation would read much better. Eg.

```clojure
(let [tx-store (assoc-in store [:db :spec] tx)
      source-set (create-source-set tx-store unit name owner-id)
      source-set-id (:id source-set)]
  (import-sources tx-store source-set-id csv-file)
  (get-source-set tx-store source-set-id owner-id))
```

You can also rename `import-sources!` and `create-source-set!` to make the side effect more visible.